### ISSUE TITLE
joint_state_publisher node issues with floating joints

### ISSUE THREAD

**Issue by shaun-edwards** [ros/robot_model#86]
System Configuration:
- ROS version: Hydro
- Ubuntu version: 12.04

The `joint_state_publisher` node crashes or produces invalid joint states when floating joints are present in the URDF.  There are two failure modes depending on what other joints are present in the URDF

Steps to reproduce:
1. URDF contains only a single floating joint
   1. URDF File:
      
      ```
      <?xml version="1.0" ?>
      <robot name = "jsp_bug" xmlns:xacro="http://ros.org/wiki/xacro">
      <link name="world" />
      <link name="floating_link" />
      <joint name="world-to-floating-link" type="floating">
      <parent link="world"/>
      <child link="floating_link"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      </joint>
      </robot>
      ```
   2.  View URDF using `roslaunch urdf_tutorial xacrodisplay.launch model:=<xacro_file>.xacro use_gui:=true`
   3.  The following error is generated by robot_state_publisher: `[ERROR] [1411791439.030568077]: Robot state publisher received an invalid joint state vector`
   4.  The `joint_state` topic echo shows:
      
      ```
      ---
      header: 
      seq: 236
      stamp: 
      secs: 1411789283
      nsecs: 295578002
      frame_id: ''
      name: ['world-to-floating-link']
      position: []
      velocity: []
      effort: []
      ---
      ```
2. URDF contains  a single floating joint and revolute joint
   1. URDF File:
      
      ```
      <?xml version="1.0" ?>
      <robot name = "jsp_bug" xmlns:xacro="http://ros.org/wiki/xacro">
      
      <link name="world" />
      <link name="floating_link" />
      <link name="revolute_link" />
      <joint name="world-to-floating-link" type="floating">
      <parent link="world"/>
      <child link="floating_link"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      </joint>
      
      <joint name="world-to-revolute-link" type="revolute">
      <origin xyz="0 0 0" rpy="0 0 0" />
      <parent link="world" />
      <child link="revolute_link" />
      <axis xyz="0 1 0" />
      <limit lower="-3.14" upper="3.14" effort="0" velocity="3.14" />
      </joint>
      </robot>
      ```
   2.  View URDF using `roslaunch urdf_tutorial xacrodisplay.launch model:=<xacro_file>.xacro use_gui:=true`
   3.  The `joint_state_publisher` node crashes with the following: 
      
      ```
      process[rviz-4]: started with pid [13283]
      Traceback (most recent call last):
      File "/opt/ros/hydro/lib/joint_state_publisher/joint_state_publisher", line 330, in <module>
      jsp.loop()
      File "/opt/ros/hydro/lib/joint_state_publisher/joint_state_publisher", line 202, in loop
      if has_position and 'position' in joint:
      TypeError: argument of type 'NoneType' is not iterable
      [joint_state_publisher-2] process has died [pid 13262, exit code 1, cmd /opt/ros/hydro/lib/joint_state_publisher/joint_state_publisher __name:=joint_state_publisher __log:=/home/ros/.ros/log/9de95ddc-45fe-11e4-b3c6-080027f3243a/joint_state_publisher-2.log].
      log file: /home/ros/.ros/log/9de95ddc-45fe-11e4-b3c6-080027f3243a/joint_state_publisher-2*.log
      ```
      
      Expected Result:
      In the first case, the floating joint is incorrectly included in joint states.  A floating joint requires a 6DOF transform, so a single joint state value is not valid.  The source for floating joints should be tf.  In the case that a valid tf is not present, joint state publisher could publish the default transform specified by `origin` in the floating joint.

**Comment by shaun-edwards** [comment]
This bug was found while trying to address this [need](http://answers.ros.org/question/191424/should-a-calibrated-frame-be-in-a-urdf/).  Essentially, I am trying to use floating joints to represent calibrated transforms.  If this had worked, I intended on writing a node that loaded calibated frames and published the appropriate tf.  The node would also support changing calibrations during runtime, allowing calibrations to be updated without restarting the ROS system.

**Comment by shaun-edwards** [comment]
I fixed the issue with the `joint_state_publisher` by treating floating joints the same as fixed joints (essentially ignoring both).  However, the issue is still not resolvable since `robot_state_publisher` (via `kdl_parser`) assumes that any unknown joint is of type KDL `Joint::None`, see [here](https://github.com/ros/robot_model/blob/indigo-devel/kdl_parser/src/kdl_parser.cpp#L67-L92).  Fixed and floating joints are published as all fixed joints.

I don't yet see a clear way to utilize floating joints in a URDF and use standards robot/joint_state_publisher nodes.

**Comment by jacquelinekay** [comment]
hey @shaun-edwards, can I close this now that #87 was merged and Ioan cherry-picked to Indigo?

**Comment by jackieokay** [comment]
I can't reproduce this error with Indigo. Was probably fixed by robot_state_publisher#32. Closing this issue.
